src/Services/CDADocumentService.php

```diff
@@ -173,21 +179,54 @@ public function generateCCDXml($pid): string
      */
     public function portalGenerateCCD($pid): string
     {
-        $url = $this->serverUrl . "/interface/modules/zend_modules/public/encounterccdadispatch";
-        $httpClient = $this->createHttpClient();
-        $response = $httpClient->request('GET', $url, [
-            'query' => [
-                'combination' => $pid,
-                'recipient' => 'patient',
-                'view' => '1',
-                'me' => session_id(),// to authenticate in CCM. Portal only.
-                'site' => $_SESSION ['site_id']
-            ]
-        ]);
+        $dispatchTable = new EncounterccdadispatchTable(new ApplicationTable());
+        $ccdaGenerator = new CcdaGenerator($dispatchTable);
+        $result = $ccdaGenerator->generate(
+            $pid,
+            null,
+            null,
+            '0',
+            '1',
+            '0',
+            null,
+            null,
+            'patient',
+            '',
+            '',
+            null,
+            []
+        );
+        $content = $result->getContent();
+        unset($result);
 
-        $status = $response->getStatusCode(); // @todo validate
+        if (str_starts_with($content, 'ERROR:')) {
+            echo "<h3>" . text($content) . "</h3>";
+            (new SystemLogger())->errorLogCaller("Error generating Portal CCDA", ['message' => $content]);
+            die();
+        }
 
-        return $response->getContent();
+        return $this->XmlToHtmlContent($content);
+    }
+
+    private function XmlToHtmlContent($content): false|string
+    {
+        $xml = simplexml_load_string($content);
+        $xsl = new DOMDocument();
+        // cda.xsl is self-contained with bootstrap and jquery.
+        $sheet = __DIR__ . '/../../interface/modules/zend_modules/public/xsl/cda.xsl';
+        $xsl->load($sheet);
+        $proc = new XSLTProcessor();
+        $proc->importStyleSheet($xsl); // attach the xsl rules
+        $outputFile = sys_get_temp_dir() . '/out_' . time() . '.html';
```

Comment:
Would it be better/more secure to use [`tempnam`](https://www.php.net/manual/en/function.tempnam.php) here?

Reply:
kind of a nit pick isn't it?:)

Reply:
@sjpadgett you don't think this might be an insecure creation of a temporary file, like https://owasp.org/www-community/vulnerabilities/Insecure_Temporary_File? tempnam would help avoid that.

Reply:
Nope